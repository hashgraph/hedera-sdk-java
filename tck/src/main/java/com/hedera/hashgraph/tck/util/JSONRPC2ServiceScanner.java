package com.hedera.hashgraph.tck.util;

import com.hedera.hashgraph.tck.annotation.JSONRPC2Service;
import com.thetransactioncompany.jsonrpc2.server.*;
import org.springframework.context.ApplicationContext;

/**
 * Utility class to register all {@link JSONRPC2Service} annotated classes
 * to the JSON-RPC dispatcher
 */
public class JSONRPC2ServiceScanner {
    private JSONRPC2ServiceScanner() {
        // private constructor for utility class
    }

    public static void registerServices(Dispatcher dispatcher, ApplicationContext context) {
        String[] serviceNames = context.getBeanNamesForAnnotation(JSONRPC2Service.class);
        if (serviceNames != null) { // NOSONAR
            for (String serviceName : serviceNames) {
                Object service = context.getBean(serviceName);
                dispatcher.register((RequestHandler) service);
            }
        }
    }
}
