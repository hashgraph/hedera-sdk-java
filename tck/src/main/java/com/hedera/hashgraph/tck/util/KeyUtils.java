package com.hedera.hashgraph.tck.util;

import com.hedera.hashgraph.sdk.Key;
import com.hedera.hashgraph.sdk.PrivateKey;
import com.hedera.hashgraph.sdk.PublicKey;

public final class KeyUtils {

    // TODO: cover with test(s)
    public static Key getKeyFromStringDER(String keyStringDER) {
        try {
            return PublicKey.fromStringDER(keyStringDER);
        } catch (Exception e) {
            return PrivateKey.fromStringDER(keyStringDER);
        }
    }
}
