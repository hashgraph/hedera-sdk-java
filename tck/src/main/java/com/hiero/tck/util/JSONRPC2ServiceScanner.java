/*-
 *
 * Hedera Java SDK
 *
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.hiero.tck.util;

import com.hiero.tck.annotation.JSONRPC2Service;
import com.thetransactioncompany.jsonrpc2.server.*;
import org.springframework.context.ApplicationContext;

/**
 * Utility class to register all {@link JSONRPC2Service} annotated classes
 * to the JSON-RPC dispatcher
 */
public class JSONRPC2ServiceScanner {
    private JSONRPC2ServiceScanner() {
        // private constructor for utility class
    }

    public static void registerServices(Dispatcher dispatcher, ApplicationContext context) {
        String[] serviceNames = context.getBeanNamesForAnnotation(JSONRPC2Service.class);
        if (serviceNames != null) { // NOSONAR
            for (String serviceName : serviceNames) {
                Object service = context.getBean(serviceName);
                dispatcher.register((RequestHandler) service);
            }
        }
    }
}
