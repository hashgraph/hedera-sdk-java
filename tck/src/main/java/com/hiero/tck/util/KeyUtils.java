/*-
 *
 * Hedera Java SDK
 *
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.hiero.tck.util;

import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.hiero.sdk.Key;
import com.hiero.sdk.PrivateKey;
import com.hiero.sdk.PublicKey;

public final class KeyUtils {

    public enum KeyType {
        ED25519_PRIVATE_KEY("ed25519PrivateKey"),
        ED25519_PUBLIC_KEY("ed25519PublicKey"),
        ECDSA_SECP256K1_PRIVATE_KEY("ecdsaSecp256k1PrivateKey"),
        ECDSA_SECP256K1_PUBLIC_KEY("ecdsaSecp256k1PublicKey"),
        LIST_KEY("keyList"),
        THRESHOLD_KEY("thresholdKey"),
        EVM_ADDRESS_KEY("evmAddress");

        private final String keyString;

        KeyType(String keyString) {
            this.keyString = keyString;
        }

        public static KeyType fromString(String keyString) {
            for (KeyType type : KeyType.values()) {
                if (type.keyString.equals(keyString)) {
                    return type;
                }
            }
            throw new IllegalArgumentException("Unknown key type: " + keyString);
        }
    }

    public static Key getKeyFromString(String keyString) throws InvalidProtocolBufferException {
        try {
            return PublicKey.fromStringDER(keyString);
        } catch (Exception e) {
            try {
                return PrivateKey.fromStringDER(keyString);
            } catch (Exception ex) {
                return Key.fromBytes(ByteString.fromHex(keyString).toByteArray());
            }
        }
    }
}
