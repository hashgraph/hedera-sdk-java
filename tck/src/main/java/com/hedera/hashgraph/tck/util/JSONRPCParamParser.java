// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.tck.util;

import com.hedera.hashgraph.tck.methods.sdk.param.CommonTransactionParams;
import com.hedera.hashgraph.tck.methods.sdk.param.CustomFee;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import net.minidev.json.JSONArray;
import net.minidev.json.JSONObject;

public class JSONRPCParamParser {

    public static Optional<CommonTransactionParams> parseCommonTransactionParams(Map<String, Object> jrpcParams)
            throws Exception {
        return parseJsonObject(jrpcParams, "commonTransactionParams", CommonTransactionParams::parse);
    }

    public static Optional<List<CustomFee>> parseCustomFees(Map<String, Object> jrpcParams) throws Exception {
        return parseJsonArray(jrpcParams, "customFees", jsonObj -> new CustomFee().parse(jsonObj));
    }

    private static <T> Optional<T> parseJsonObject(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> parser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONObject jsonObject = (JSONObject) params.get(key);
        return Optional.of(parser.apply(jsonObject));
    }

    private static <T> Optional<List<T>> parseJsonArray(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> elementParser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONArray jsonArray = (JSONArray) params.get(key);
        List<T> results = new ArrayList<>();

        for (Object o : jsonArray) {
            JSONObject jsonObject = (JSONObject) o;
            results.add(elementParser.apply(jsonObject));
        }
        return Optional.of(results);
    }

    @FunctionalInterface
    private interface ThrowingFunction<T, R> {
        R apply(T t) throws Exception;
    }
}
