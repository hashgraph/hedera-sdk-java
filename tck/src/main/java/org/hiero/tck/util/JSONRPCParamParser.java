// SPDX-License-Identifier: Apache-2.0
package org.hiero.tck.util;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import net.minidev.json.JSONArray;
import net.minidev.json.JSONObject;
import org.hiero.tck.exception.JSONRPCParseException;
import org.hiero.tck.methods.sdk.param.CommonTransactionParams;
import org.hiero.tck.methods.sdk.param.CustomFee;

public class JSONRPCParamParser {

    public static Optional<CommonTransactionParams> parseCommonTransactionParams(Map<String, Object> jrpcParams) {
        try {
            return parseJsonObject(jrpcParams, "commonTransactionParams", CommonTransactionParams::parse);
        } catch (Exception e) {
            throw new JSONRPCParseException("Failed to parse CommonTransactionParams", e);
        }
    }

    public static Optional<List<CustomFee>> parseCustomFees(Map<String, Object> jrpcParams) {
        try {
            return parseJsonArray(jrpcParams, "customFees", jsonObj -> {
                try {
                    return new CustomFee().parse(jsonObj);
                } catch (Exception e) {
                    throw new JSONRPCParseException("Failed to parse CustomFee", e);
                }
            });
        } catch (Exception e) {
            throw new JSONRPCParseException("Failed to parse CustomFees array", e);
        }
    }

    private static <T> Optional<T> parseJsonObject(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> parser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONObject jsonObject = (JSONObject) params.get(key);
        return Optional.of(parser.apply(jsonObject));
    }

    private static <T> Optional<List<T>> parseJsonArray(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> elementParser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONArray jsonArray = (JSONArray) params.get(key);
        List<T> results = new ArrayList<>();

        for (Object o : jsonArray) {
            JSONObject jsonObject = (JSONObject) o;
            results.add(elementParser.apply(jsonObject));
        }
        return Optional.of(results);
    }

    @FunctionalInterface
    private interface ThrowingFunction<T, R> {
        R apply(T t) throws Exception;
    }
}
