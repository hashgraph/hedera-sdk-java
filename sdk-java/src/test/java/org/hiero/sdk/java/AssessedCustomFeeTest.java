// SPDX-License-Identifier: Apache-2.0
package org.hiero.sdk.java;

import static org.assertj.core.api.Assertions.assertThat;

import io.github.jsonSnapshot.SnapshotMatcher;
import java.util.List;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class AssessedCustomFeeTest {

    private static final int amount = 1;
    private static final TokenId tokenId = new TokenId(2, 3, 4);
    private static final AccountId feeCollector = new AccountId(5, 6, 7);
    private static final List<AccountId> payerAccountIds =
            List.of(new AccountId(8, 9, 10), new AccountId(11, 12, 13), new AccountId(14, 15, 16));

    private final org.hiero.sdk.java.proto.AssessedCustomFee fee =
            org.hiero.sdk.java.proto.AssessedCustomFee.newBuilder()
                    .setAmount(amount)
                    .setTokenId(tokenId.toProtobuf())
                    .setFeeCollectorAccountId(feeCollector.toProtobuf())
                    .addAllEffectivePayerAccountId(
                            payerAccountIds.stream().map(AccountId::toProtobuf).toList())
                    .build();

    @BeforeAll
    public static void beforeAll() {
        SnapshotMatcher.start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        SnapshotMatcher.validateSnapshots();
    }

    AssessedCustomFee spawnAssessedCustomFeeExample() {
        return new AssessedCustomFee(
                201,
                TokenId.fromString("1.2.3"),
                AccountId.fromString("4.5.6"),
                List.of(AccountId.fromString("0.0.1"), AccountId.fromString("0.0.2"), AccountId.fromString("0.0.3")));
    }

    @Test
    void shouldSerialize() throws Exception {
        var originalAssessedCustomFee = spawnAssessedCustomFeeExample();
        byte[] assessedCustomFeeBytes = originalAssessedCustomFee.toBytes();
        var copyAssessedCustomFee = AssessedCustomFee.fromBytes(assessedCustomFeeBytes);
        assertThat(originalAssessedCustomFee.toString().replaceAll("@[A-Za-z0-9]+", ""))
                .isEqualTo(copyAssessedCustomFee.toString().replaceAll("@[A-Za-z0-9]+", ""));
        SnapshotMatcher.expect(originalAssessedCustomFee.toString().replaceAll("@[A-Za-z0-9]+", ""))
                .toMatchSnapshot();
    }

    @Test
    void fromProtobuf() {
        SnapshotMatcher.expect(AssessedCustomFee.fromProtobuf(fee).toString()).toMatchSnapshot();
    }

    @Test
    void toProtobuf() {
        SnapshotMatcher.expect(AssessedCustomFee.fromProtobuf(fee).toProtobuf().toString())
                .toMatchSnapshot();
    }

    @Test
    void shouldBytes() throws Exception {
        var assessedCustomFee = spawnAssessedCustomFeeExample();
        var tx2 = AssessedCustomFee.fromBytes(assessedCustomFee.toBytes());
        assertThat(tx2).hasToString(assessedCustomFee.toString());
    }
}
