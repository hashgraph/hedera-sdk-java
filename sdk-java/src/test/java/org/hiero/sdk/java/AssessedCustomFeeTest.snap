org.hiero.sdk.java.AssessedCustomFeeTest.fromProtobuf=[
  "AssessedCustomFee{amount=1, tokenId=2.3.4, feeCollectorAccountId=5.6.7, payerAccountIdList=[8.9.10, 11.12.13, 14.15.16]}"
]


org.hiero.sdk.java.AssessedCustomFeeTest.shouldSerialize=[
  "AssessedCustomFee{amount=201, tokenId=1.2.3, feeCollectorAccountId=4.5.6, payerAccountIdList=[0.0.1, 0.0.2, 0.0.3]}"
]


org.hiero.sdk.java.AssessedCustomFeeTest.toProtobuf=[
  "# org.hiero.sdk.java.proto.AssessedCustomFee@55ba505b\namount: 1\neffective_payer_account_id {\n  account_num: 10\n  realm_num: 9\n  shard_num: 8\n}\neffective_payer_account_id {\n  account_num: 13\n  realm_num: 12\n  shard_num: 11\n}\neffective_payer_account_id {\n  account_num: 16\n  realm_num: 15\n  shard_num: 14\n}\nfee_collector_account_id {\n  account_num: 7\n  realm_num: 6\n  shard_num: 5\n}\ntoken_id {\n  realm_num: 3\n  shard_num: 2\n  token_num: 4\n}"
]
